File: backend/app/utils/production_logger.py
"""
Production Logger
Optimized logging system for production environments with rate limiting
and intelligent log filtering to prevent Railway rate limit issues.
"""

import logging
import time
import os
from typing import Dict, Any, Optional
from collections import defaultdict, deque
from threading import Lock
import json

class ProductionLogger:
    """
    Production-optimized logger with rate limiting and intelligent filtering.
    """
    
    def __init__(self, name: str, level: str = 'WARNING'):
        self.name = name
        self.logger = logging.getLogger(name)
        self.logger.setLevel(getattr(logging, level.upper()))
        
        # Rate limiting configuration
        self.rate_limits = {
            'cursor_move': {'interval': 30.0, 'max_per_interval': 5},
            'object_update': {'interval': 10.0, 'max_per_interval': 10},
            'auth': {'interval': 60.0, 'max_per_interval': 3},
            'ai_request': {'interval': 30.0, 'max_per_interval': 5},
            'network_check': {'interval': 60.0, 'max_per_interval': 2},
            'error': {'interval': 0.0, 'max_per_interval': 100},  # Always log errors
            'info': {'interval': 5.0, 'max_per_interval': 3},
            'debug': {'interval': 0.0, 'max_per_interval': 0}  # Disabled in production
        }
        
        # Track log counts per interval
        self.log_counts: Dict[str, deque] = defaultdict(lambda: deque())
        self.lock = Lock()
        
        # Environment-based configuration
        self.is_production = os.environ.get('FLASK_ENV') == 'production'
        self.is_railway = os.environ.get('RAILWAY_ENVIRONMENT') is not None
        
        # Aggregation for high-frequency events
        self.aggregated_logs: Dict[str, Dict[str, Any]] = {}
        self.aggregation_interval = 30.0  # seconds
        self.last_aggregation = time.time()
    
    def _should_log(self, event_type: str) -> bool:
        """Check if we should log this event based on rate limiting."""
        if not self.is_production:
            return True  # No rate limiting in development
        
        with self.lock:
            now = time.time()
            rate_config = self.rate_limits.get(event_type, self.rate_limits['info'])
            
            # Clean old entries
            cutoff_time = now - rate_config['interval']
            while (self.log_counts[event_type] and 
                   self.log_counts[event_type][0] < cutoff_time):
                self.log_counts[event_type].popleft()
            
            # Check if we can log
            if len(self.log_counts[event_type]) < rate_config['max_per_interval']:
                self.log_counts[event_type].append(now)
                return True
            
            return False
    
    def _aggregate_log(self, event_type: str, message: str, level: str = 'INFO'):
        """Aggregate similar logs to reduce volume."""
        if not self.is_production:
            return  # No aggregation in development
        
        now = time.time()
        key = f"{event_type}_{level}"
        
        if key not in self.aggregated_logs:
            self.aggregated_logs[key] = {
                'count': 0,
                'first_seen': now,
                'last_seen': now,
                'message': message,
                'level': level
            }
        
        self.aggregated_logs[key]['count'] += 1
        self.aggregated_logs[key]['last_seen'] = now
        
        # Flush aggregated logs periodically
        if now - self.last_aggregation > self.aggregation_interval:
            self._flush_aggregated_logs()
            self.last_aggregation = now
    
    def _flush_aggregated_logs(self):
        """Flush aggregated logs to reduce volume."""
        for key, data in self.aggregated_logs.items():
            if data['count'] > 1:
                duration = data['last_seen'] - data['first_seen']
                message = f"[AGGREGATED] {data['message']} (occurred {data['count']} times in {duration:.1f}s)"
                
                if data['level'] == 'ERROR':
                    self.logger.error(message)
                elif data['level'] == 'WARNING':
                    self.logger.warning(message)
                else:
                    self.logger.info(message)
            elif data['count'] == 1:
                # Single occurrence, log normally
                if data['level'] == 'ERROR':
                    self.logger.error(data['message'])
                elif data['level'] == 'WARNING':
                    self.logger.warning(data['message'])
                else:
                    self.logger.info(data['message'])
        
        self.aggregated_logs.clear()
    
    def log_cursor_move(self, user_id: str, position: Dict[str, float]):
        """Log cursor movement with heavy rate limiting."""
        if self._should_log('cursor_move'):
            self._aggregate_log('cursor_move', f"Cursor: {user_id} -> ({position['x']:.1f}, {position['y']:.1f})")
    
    def log_object_update(self, user_id: str, object_id: str, action: str):
        """Log object updates with rate limiting."""
        if self._should_log('object_update'):
            self._aggregate_log('object_update', f"Object {action}: {object_id} by {user_id}")
    
    def log_auth(self, user_id: str, action: str):
        """Log authentication events with rate limiting."""
        if self._should_log('auth'):
            self.logger.info(f"Auth: {action} for user {user_id}")
    
    def log_ai_request(self, user_id: str, request_type: str, success: bool = True):
        """Log AI requests with rate limiting."""
        if self._should_log('ai_request'):
            status = "SUCCESS" if success else "FAILED"
            self._aggregate_log('ai_request', f"AI {request_type}: {status} for user {user_id}")
    
    def log_network_check(self, service: str, status: str):
        """Log network health checks with heavy rate limiting."""
        if self._should_log('network_check'):
            self._aggregate_log('network_check', f"Network check: {service} -> {status}")
    
    def log_error(self, message: str, error: Exception = None, context: Dict[str, Any] = None):
        """Always log errors (highest priority)."""
        if error:
            error_msg = f"ERROR: {message} - {str(error)}"
        else:
            error_msg = f"ERROR: {message}"
        
        if context:
            error_msg += f" | Context: {json.dumps(context, default=str)}"
        
        self.logger.error(error_msg)
    
    def log_warning(self, message: str, context: Dict[str, Any] = None):
        """Log warnings with rate limiting."""
        if self._should_log('info'):
            warning_msg = f"WARNING: {message}"
            if context:
                warning_msg += f" | Context: {json.dumps(context, default=str)}"
            self.logger.warning(warning_msg)
    
    def log_info(self, message: str, context: Dict[str, Any] = None):
        """Log info with rate limiting."""
        if self._should_log('info'):
            info_msg = f"INFO: {message}"
            if context:
                info_msg += f" | Context: {json.dumps(context, default=str)}"
            self.logger.info(info_msg)
    
    def log_debug(self, message: str, context: Dict[str, Any] = None):
        """Log debug messages (disabled in production)."""
        if not self.is_production and self._should_log('debug'):
            debug_msg = f"DEBUG: {message}"
            if context:
                debug_msg += f" | Context: {json.dumps(context, default=str)}"
            self.logger.debug(debug_msg)
    
    def log_performance(self, operation: str, duration: float, user_id: str = None):
        """Log performance metrics with rate limiting."""
        if self._should_log('info'):
            perf_msg = f"PERF: {operation} took {duration:.3f}s"
            if user_id:
                perf_msg += f" for user {user_id}"
            self.logger.info(perf_msg)
    
    def get_log_stats(self) -> Dict[str, Any]:
        """Get current logging statistics."""
        with self.lock:
            stats = {}
            for event_type, times in self.log_counts.items():
                stats[event_type] = {
                    'count': len(times),
                    'rate_limit': self.rate_limits.get(event_type, {}),
                    'last_log': times[-1] if times else None
                }
            return stats
    
    def flush_aggregated(self):
        """Manually flush aggregated logs."""
        self._flush_aggregated_logs()


# Global production logger instance
production_logger = ProductionLogger('collabcanvas_production', 'WARNING')

# Convenience functions
def log_cursor_move(user_id: str, position: Dict[str, float]):
    production_logger.log_cursor_move(user_id, position)

def log_object_update(user_id: str, object_id: str, action: str):
    production_logger.log_object_update(user_id, object_id, action)

def log_auth(user_id: str, action: str):
    production_logger.log_auth(user_id, action)

def log_ai_request(user_id: str, request_type: str, success: bool = True):
    production_logger.log_ai_request(user_id, request_type, success)

def log_network_check(service: str, status: str):
    production_logger.log_network_check(service, status)

def log_error(message: str, error: Exception = None, context: Dict[str, Any] = None):
    production_logger.log_error(message, error, context)

def log_warning(message: str, context: Dict[str, Any] = None):
    production_logger.log_warning(message, context)

def log_info(message: str, context: Dict[str, Any] = None):
    production_logger.log_info(message, context)

def log_debug(message: str, context: Dict[str, Any] = None):
    production_logger.log_debug(message, context)

def log_performance(operation: str, duration: float, user_id: str = None):
    production_logger.log_performance(operation, duration, user_id)

def get_log_stats() -> Dict[str, Any]:
    return production_logger.get_log_stats()

def flush_aggregated_logs():
    production_logger.flush_aggregated()

File: frontend/src/utils/productionLogger.ts
/**
 * Production Logger
 * Optimized logging system for production environments
 * Reduces console.log spam and implements rate limiting
 */

interface LogEntry {
  timestamp: number
  level: string
  message: string
  context?: Record<string, unknown>
}

interface RateLimitConfig {
  interval: number // milliseconds
  maxLogs: number
}

class ProductionLogger {
  private isProduction: boolean
  private isDevelopment: boolean
  private logBuffer: LogEntry[] = []
  private rateLimits: Map<string, { count: number; resetTime: number }> = new Map()
  private maxBufferSize = 100
  private flushInterval = 30000 // 30 seconds

  // Rate limiting configuration
  private rateLimitConfigs: Record<string, RateLimitConfig> = {
    'network_check': { interval: 60000, maxLogs: 2 }, // 2 logs per minute
    'cursor_move': { interval: 30000, maxLogs: 5 },   // 5 logs per 30 seconds
    'object_update': { interval: 10000, maxLogs: 10 }, // 10 logs per 10 seconds
    'socket_event': { interval: 5000, maxLogs: 20 },   // 20 logs per 5 seconds
    'error': { interval: 0, maxLogs: 100 },            // Always log errors
    'warning': { interval: 10000, maxLogs: 10 },       // 10 warnings per 10 seconds
    'info': { interval: 5000, maxLogs: 5 },            // 5 info logs per 5 seconds
    'debug': { interval: 0, maxLogs: 0 }               // Disabled in production
  }

  constructor() {
    this.isProduction = import.meta.env.PROD || import.meta.env.MODE === 'production'
    this.isDevelopment = import.meta.env.DEV || import.meta.env.MODE === 'development'
    
    // Start periodic buffer flush
    if (this.isProduction) {
      setInterval(() => this.flushBuffer(), this.flushInterval)
    }
  }

  private shouldLog(category: string): boolean {
    if (this.isDevelopment) {
      return true // No rate limiting in development
    }

    const config = this.rateLimitConfigs[category] || this.rateLimitConfigs['info']
    const now = Date.now()
    const rateLimit = this.rateLimits.get(category)

    if (!rateLimit || now > rateLimit.resetTime) {
      // Reset or initialize rate limit
      this.rateLimits.set(category, {
        count: 1,
        resetTime: now + config.interval
      })
      return true
    }

    if (rateLimit.count < config.maxLogs) {
      rateLimit.count++
      return true
    }

    return false
  }

  private addToBuffer(level: string, message: string, context?: Record<string, unknown>): void {
    if (this.isDevelopment) {
      return // No buffering in development
    }

    this.logBuffer.push({
      timestamp: Date.now(),
      level,
      message,
      context
    })

    // Prevent buffer overflow
    if (this.logBuffer.length > this.maxBufferSize) {
      this.logBuffer = this.logBuffer.slice(-this.maxBufferSize)
    }
  }

  private flushBuffer(): void {
    if (this.logBuffer.length === 0) {
      return
    }

    // Group logs by level and message pattern
    const groupedLogs = new Map<string, { count: number; firstSeen: number; lastSeen: number; level: string; message: string }>()

    for (const log of this.logBuffer) {
      const key = `${log.level}:${log.message}`
      if (groupedLogs.has(key)) {
        const existing = groupedLogs.get(key)!
        existing.count++
        existing.lastSeen = log.timestamp
      } else {
        groupedLogs.set(key, {
          count: 1,
          firstSeen: log.timestamp,
          lastSeen: log.timestamp,
          level: log.level,
          message: log.message
        })
      }
    }

    // Log aggregated results
    for (const [_key, data] of groupedLogs) {
      if (data.count > 1) {
        const duration = (data.lastSeen - data.firstSeen) / 1000
        const aggregatedMessage = `[AGGREGATED] ${data.message} (occurred ${data.count} times in ${duration.toFixed(1)}s)`
        this.logToConsole(data.level, aggregatedMessage)
      } else {
        this.logToConsole(data.level, data.message)
      }
    }

    this.logBuffer = []
  }

  private logToConsole(level: string, message: string, context?: Record<string, unknown>): void {
    const timestamp = new Date().toISOString()
    const logMessage = `[${timestamp}] ${level.toUpperCase()}: ${message}`
    
    if (context) {
      const contextStr = typeof context === 'object' ? JSON.stringify(context, null, 2) : String(context)
      console.log(logMessage, contextStr)
    } else {
      console.log(logMessage)
    }
  }

  // Public logging methods
  error(message: string, context?: Record<string, unknown>): void {
    if (this.shouldLog('error')) {
      this.logToConsole('error', message, context)
    } else {
      this.addToBuffer('error', message, context)
    }
  }

  warning(message: string, context?: Record<string, unknown>): void {
    if (this.shouldLog('warning')) {
      this.logToConsole('warning', message, context)
    } else {
      this.addToBuffer('warning', message, context)
    }
  }

  info(message: string, context?: Record<string, unknown>): void {
    if (this.shouldLog('info')) {
      this.logToConsole('info', message, context)
    } else {
      this.addToBuffer('info', message, context)
    }
  }

  debug(message: string, context?: Record<string, unknown>): void {
    if (this.isDevelopment && this.shouldLog('debug')) {
      this.logToConsole('debug', message, context)
    }
  }

  // Specialized logging methods
  networkCheck(service: string, status: string): void {
    if (this.shouldLog('network_check')) {
      this.logToConsole('info', `Network check: ${service} -> ${status}`)
    } else {
      this.addToBuffer('info', `Network check: ${service} -> ${status}`)
    }
  }

  cursorMove(userId: string, position: { x: number; y: number }): void {
    if (this.shouldLog('cursor_move')) {
      this.logToConsole('debug', `Cursor: ${userId} -> (${position.x.toFixed(1)}, ${position.y.toFixed(1)})`)
    } else {
      this.addToBuffer('debug', `Cursor: ${userId} -> (${position.x.toFixed(1)}, ${position.y.toFixed(1)})`)
    }
  }

  objectUpdate(userId: string, objectId: string, action: string): void {
    if (this.shouldLog('object_update')) {
      this.logToConsole('debug', `Object ${action}: ${objectId} by ${userId}`)
    } else {
      this.addToBuffer('debug', `Object ${action}: ${objectId} by ${userId}`)
    }
  }

  socketEvent(eventType: string, success: boolean, userId?: string): void {
    if (this.shouldLog('socket_event')) {
      const status = success ? 'SUCCESS' : 'FAILED'
      const userInfo = userId ? ` for user ${userId}` : ''
      this.logToConsole('info', `Socket event: ${eventType} -> ${status}${userInfo}`)
    } else {
      const status = success ? 'SUCCESS' : 'FAILED'
      const userInfo = userId ? ` for user ${userId}` : ''
      this.addToBuffer('info', `Socket event: ${eventType} -> ${status}${userInfo}`)
    }
  }

  // Performance logging
  performance(operation: string, duration: number, userId?: string): void {
    if (this.shouldLog('info')) {
      const userInfo = userId ? ` for user ${userId}` : ''
      this.logToConsole('info', `PERF: ${operation} took ${duration.toFixed(3)}s${userInfo}`)
    } else {
      const userInfo = userId ? ` for user ${userId}` : ''
      this.addToBuffer('info', `PERF: ${operation} took ${duration.toFixed(3)}s${userInfo}`)
    }
  }

  // Manual buffer flush
  flush(): void {
    this.flushBuffer()
  }

  // Get current buffer status
  getBufferStatus(): { size: number; oldestEntry?: number; newestEntry?: number } {
    if (this.logBuffer.length === 0) {
      return { size: 0 }
    }

    return {
      size: this.logBuffer.length,
      oldestEntry: this.logBuffer[0].timestamp,
      newestEntry: this.logBuffer[this.logBuffer.length - 1].timestamp
    }
  }

  // Get rate limit status
  getRateLimitStatus(): Record<string, { count: number; resetTime: number; maxLogs: number }> {
    const status: Record<string, { count: number; resetTime: number; maxLogs: number }> = {}
    
    for (const [category, rateLimit] of this.rateLimits) {
      const config = this.rateLimitConfigs[category] || this.rateLimitConfigs['info']
      status[category] = {
        count: rateLimit.count,
        resetTime: rateLimit.resetTime,
        maxLogs: config.maxLogs
      }
    }
    
    return status
  }
}

// Create singleton instance
export const productionLogger = new ProductionLogger()

// Export convenience functions
export const logError = (message: string, context?: Record<string, unknown>) => productionLogger.error(message, context)
export const logWarning = (message: string, context?: Record<string, unknown>) => productionLogger.warning(message, context)
export const logInfo = (message: string, context?: Record<string, unknown>) => productionLogger.info(message, context)
export const logDebug = (message: string, context?: Record<string, unknown>) => productionLogger.debug(message, context)
export const logNetworkCheck = (service: string, status: string) => productionLogger.networkCheck(service, status)
export const logCursorMove = (userId: string, position: { x: number; y: number }) => productionLogger.cursorMove(userId, position)
export const logObjectUpdate = (userId: string, objectId: string, action: string) => productionLogger.objectUpdate(userId, objectId, action)
export const logSocketEvent = (eventType: string, success: boolean, userId?: string) => productionLogger.socketEvent(eventType, success, userId)
export const logPerformance = (operation: string, duration: number, userId?: string) => productionLogger.performance(operation, duration, userId)
export const flushLogs = () => productionLogger.flush()
export const getLogStatus = () => ({
  buffer: productionLogger.getBufferStatus(),
  rateLimits: productionLogger.getRateLimitStatus()
})

export default productionLogger
